salt/loader.py

```diff
@@ -745,10 +742,12 @@ def grains(opts, force_refresh=False, proxy=None):
         ret = funcs[key]()
         if not isinstance(ret, dict):
             continue
-        if is_filter:
-            bkeys = set(ret.keys()).intersection(blist)
-            for bkey in bkeys:
-                del ret[bkey]
+        if blist:
+            for key in ret.keys():
```

Comment:
This will break on Python 3:

```python
>>> x = {'foo': 'bar', 'abc': 123}
>>> for key in x.keys():
...     del x[key]
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: dictionary changed size during iteration
>>>
```

This is because on Python 3, `dict.keys()` is a view and not a separate list like in Python 2.

To fix this, use `list(ret)` instead of `ret.keys()`.